Collect statistics from pg_stat_io (Postgres 16+)

seanlinsley opened the pull request:
https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STAT-IO-VIEW

In a future PR we'll want to handle `stats_reset`, but that's been skipped for now.

lfittl commented on input/postgres/server_stats.go:
```diff
@@ -69,8 +69,23 @@ COALESCE((
+	   evictions,
+	   reuses,
+	   fsyncs
+  FROM pg_stat_io
```
@seanlinsley We should probably add the `_time` fields as well here - AFAIK they didn't exist yet when I first prototyped this, but were added later (and are very useful to understand whether its likely the OS actually issued an I/O request to disk, vs getting something from the page cache).

We could probably also add `writebacks` and `writeback_time` (https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STAT-IO-VIEW) - its a bit specialized, but potentially useful.

lfittl commented on runner/diff.go:
```diff
@@ -195,3 +196,25 @@ func diffDatabaseStats(new state.PostgresDatabaseStatsMap, prev state.PostgresDa
+		}
+		// Skip over empty diffs (which can occur either because there was no activity, or for fixed entries that never saw activity)
+		if s.Reads != 0 || s.Writes != 0 || s.Extends != 0 ||
+			s.Evictions != 0 || s.Reuses != 0 || s.Fsyncs != 0 {
```
Not a big deal, but we could add Writebacks and `Hits` here as well (it seems very unlikely that that'd be the only activity on a system, but easier to be consistent here for a future reader of the code).

lfittl reviewed (commented):
Looks good!

seanlinsley commented:
Changes since the previous iteration of this PR:
- `backend_type` is no longer a string, instead it reuses the `BackendCountStatistic.BackendType` enum
- `io_object` and `io_context` are no longer strings, instead using new enums
- The protobuf struct is longer prefixed with `Historic`. AFAIK that's a namign pattern used to distinguish between the many Go structs needed for diffing, but there's no reason to expose that internal complexity to the protobuf structure.
- `collected_interval_secs` has been renamed to `collected_secs`. I feel that communicates the same intent with fewer words, and I personally find "interval" confusing in this context.

lfittl commented on output/transform/postgres.go:
```diff
@@ -167,6 +170,83 @@ func transformPostgresServerStats(s snapshot.FullSnapshot, newState state.Persis
+			case "walsender":
+				stat.BackendType = snapshot.BackendCountStatistic_WALSENDER
+			case "walwriter":
+				stat.BackendType = snapshot.BackendCountStatistic_WALWRITER
```
We're missing the "slotsync worker" backend type here, that was added in Postgres 17: https://www.postgresql.org/docs/current/logicaldecoding-explanation.html#LOGICALDECODING-REPLICATION-SLOTS-SYNCHRONIZATION

(I think to be able to account for all I/O its helpful to include it, even if an edge case)

lfittl commented on runner/full_1min.go:
```diff
@@ -12,7 +12,7 @@ import (
 )
 
-func gatherQueryStatsForServer(ctx context.Context, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) (state.PersistedState, error) {
+func gather1minStatsForServer(ctx context.Context, server *state.Server, globalCollectionOpts state.CollectionOpts, logger *util.Logger) (state.PersistedState, error) {
```
The `1min` in the function name jumped out a bit at me at first glance, but not sure there is a better name here - using something like "high frequency" in the name is probably more confusing than helpful.

lfittl commented on runner/full_1min.go:
```diff
@@ -51,33 +51,45 @@ func gatherQueryStatsForServer(ctx context.Context, server *state.Server, global
 
-	diffedStatementStats := diffStatements(newState.StatementStats, server.PrevState.StatementStats)
-	collectedIntervalSecs := uint32(newState.LastStatementStatsAt.Sub(server.PrevState.LastStatementStatsAt) / time.Second)
+	timeKey := state.PostgresStatementStatsTimeKey{
```
I wonder if we should rename the `PostgresStatementStatsTimeKey` struct to `HistoricStatsTimeKey` or something, so its clear its not specific to statements?

lfittl reviewed (commented):
Overall looks good - I want to do one more test together with the storage PR, but left a few minor notes in the meantime.

lfittl reviewed (approved):
:+1:

lfittl reviewed (approved):
:+1:
